### PR TITLE
Prevent duplicate names for vents and scrubbers in the air alarm

### PR DESCRIPTION
**Pull request by nightred**
## About The Pull Request

It is always a pain adding in a new vent or scrubber to an area and suddenly you have duplicate names.
This prevents duplicate names for vents and scrubbers in an area.

## Why It's Good For The Game

Makes engineering's life better, duplicate names are bad.

## Changelog
:cl:
fix: Prevent duplicate vent and scrubber names in the air alarm interface.
/:cl:

### CODE REVIEW

**Comment by antropod** [comment]
The only issue I see is quadratic worst case time complexity of this algorithm
If you have 100 scrubbers, named #100 to #199 you have to do 10000 checks to find next free name.
But maybe its not an issue since we usually don't have this many scrubbers in the area

**Comment by TemporalOroboros** [comment]
Two words, gas farms.

**Comment by nightred** [comment]
> The only issue I see is quadratic worst case time complexity of this algorithm
> If you have 100 scrubbers, named #100 to #199 you have to do 10000 checks to find next free name.
> But maybe its not an issue since we usually don't have this many scrubbers in the area

Ok, I changed this, 100 scrubbers should now take 100 checks to validate at the worse.
Technically still a lot of checks using to validate if name in list but this works better in my opinion.

**Comment by antropod** [comment]
em, I think this is still quadratic.
one search in list is O(n)
and you need to do that n times in worst case
so you get O(n^2) time complexity

**Comment by antropod** [comment]
This is what I was thinking as a solution for this problem, which has O(n) time complexity
Algorithm is expressed in pseudo-code
```
# By "name" here I mean sequential number of the scrubber/vent in the area
# I keep track of all used names in the area with an array `names`
# Array index means name, array value means tag_id
# If value is null, the name is not used
#
# So for example area has vents #1, #2 and #4, then array of names looks like
# ["tag1", "tag2", null, "tag4"]
#
# When you allocate new name for tag_id, you scan array for free slot,
# assign tag_id to that slot and return its index
# if all slots occupied, you add value to the end of array
#
# When you deallocate name for tag_id, you find tag_id in the array
# and set value of that element to null (or to some other meaningless value for tag_id)

# names = []

func allocate_name(names, tag_id):
    for i = 1..names.length:
        if names[i] == tag_id:
            # this tag_id already has a name
            return i
        if names[i] == null:
            # found empty slot in the array
            names[i] = tag_id
            return i
    # end of array, add new value to the end
    names.append(tag_id)
    return names.length

func deallocate_name(names, tag_id):
    # function `find` finds index of element in the array, and returns 0 if not found
    i = names.find(tag_id) 
    if (i == 0):
        # name for this tag was never allocated
        return false
    # erase tag_id value
    names[i] = null
    return true
```

**Comment by nightred** [comment]
This uses a style like was suggested.

**Comment by antropod** [comment]
All scrubbers and vents named by their default name
you can see it by hovering mouse over scrubber or vent, its named just "air scrubber", I think you missed `name = ...`

Also before this commit name was assigned in `broadcast_status()`, are you sure you want to move it to `New()`? I don't know the reason for that so I wouldn't change it.

Also I think its good idea to move all duplicated code into function, for example inside of `area` class
One function for allocating new name, one for deallocating

**Comment by antropod** [comment]
Also, ordering in the interface getting messed up
image

**Comment by antropod** [comment]
Other than that, looks good.
I removed scrubber and put it back, name was not duplicated, it reused old name.
Same with removing multiple scrubbers, names getting assigned in order.
New scrubbers get new name.

**Comment by nightred** [comment]
Vents in numbered order now when listed in air alarm.
Naming of the air alarms for mouse over.

**Comment by antropod** [comment]
Looking good!
Everything works

**Comment by optimumtact** [comment]
this is kind of silly, just have a global map of
`name -> latest id tag `


and when you get the tag id, just use the name to look in the map to determine if it exists before or not
```
#need to handle if there is no previous mapping for this name and start the tagid at 1.
tagid = map[name]
# update the mapping
map[name] = tagid+1
```

Now it's an O(1) operation and you can never have any collision, and you only pay the cost of mapping in use names.

Then automate all tag generation and don't let mappers set it and you're done.

**Comment by antropod** [comment]
I think you misunderstood problem.

Unary device does not know it's name, it has name "air scrubber" or "air vent" by default. So you can't look up `tagid` by name in the global map.
And there's no point of doing so, `tagid` (or `id_tag` in the actual code) is already assigned uniquely to each new unary deivce by using global counter which increments each time new device is created.

What are we solving here is numbering vents in the area.
So for example, if you have 3 vents in the area named #1, #2 and #3, and then you remove vent #2 and put it back, you will have vents with names #1, #3, #3
This happens because new name is assigned by formula `A.air_scrub_names.len + 1`.

Suggested solution fixes this problem by finding free name with lowest number in the area, so all vents and scrubbers have nice sequential names with no duplicates.

As side note, you can solve this problem in O(1) if you just use autoincrement counter for each area, but it feels weird ingame and counterintuitive. For example, when you unscrew and screw device back, you will get larger and larger values in its name, and list of vents and scrubbers in the area will have gaps. It seems like minor issue, but it's actually even worse than having duplicate names.

**Comment by optimumtact** [comment]
so do a map of that instead, and just increment it like the fukken tag id does anyway

**Comment by antropod** [comment]
We have good solution, which gives you nice names and reuses gaps.
This is not critical path which is executed 20 times a second, also I'm not certain that BYOND maps are O(1) lookup time (they're more like lists of key-value pairs, not real hash maps).
And current solution is better in terms of user experience.

**Comment by nightred** [comment]
I will rework this, and do a simple mapping without the ordered numbers, it might be tags or words but no more numbers.

**Comment by antropod** [comment]
That's cool solution

File: code/game/machinery/_machinery.dm [hunk 1]
```diff
@@ -601,3 +601,25 @@ Class Procs:
...
+	// machine id's should be fun random chars hinting at a larger world
+	for(var/i = 1 to len)
+		switch(rand(1,3))
+			if(1)
+				new_name += ascii2text(rand(65, 90)) // A - Z
+			if(2)
+				new_name += ascii2text(rand(97,122)) // a - z
+			if(3)
+				new_name += ascii2text(rand(48, 57)) // 0 - 9
+	return new_name
```
Comment (Rohesie) [inline code/game/machinery/_machinery.dm:625]:
```suggestion
	var/list/new_name = list()
	// machine id's should be fun random chars hinting at a larger world
	for(var/i = 1 to len)
		switch(rand(1,3))
			if(1)
				new_name += ascii2text(rand(65, 90)) // A - Z
			if(2)
				new_name += ascii2text(rand(97,122)) // a - z
			if(3)
				new_name += ascii2text(rand(48, 57)) // 0 - 9
	return new_name.Join()
```

**Pull request merged**